<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Проверка-на-сбалансированность-данных" data-toc-modified-id="Проверка-на-сбалансированность-данных-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Проверка на сбалансированность данных</a></span></li><li><span><a href="#Построение-моделей-на-несбалансированных-данных" data-toc-modified-id="Построение-моделей-на-несбалансированных-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Построение моделей на несбалансированных данных</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Модель-случайного-леса-с-подбором-параметров" data-toc-modified-id="Модель-случайного-леса-с-подбором-параметров-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Модель случайного леса с подбором параметров</a></span></li></ul></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Изменение-порога" data-toc-modified-id="Изменение-порога-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Изменение порога</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Модель-случайного-леса-с-подбором-параметров" data-toc-modified-id="Модель-случайного-леса-с-подбором-параметров-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Модель случайного леса с подбором параметров</a></span></li></ul></li><li><span><a href="#Балансировка-весов-класса" data-toc-modified-id="Балансировка-весов-класса-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Балансировка весов класса</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Случайный-лес-с-подбором-гиперпараметров" data-toc-modified-id="Случайный-лес-с-подбором-гиперпараметров-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Случайный лес с подбором гиперпараметров</a></span></li></ul></li><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Увеличение выборки</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Случайный-лес-с-подбором-гиперпараметров" data-toc-modified-id="Случайный-лес-с-подбором-гиперпараметров-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Случайный лес с подбором гиперпараметров</a></span></li></ul></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Уменьшение выборки</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Случайный-лес-с-подбором-гиперпараметров" data-toc-modified-id="Случайный-лес-с-подбором-гиперпараметров-3.4.2"><span class="toc-item-num">3.4.2&nbsp;&nbsp;</span>Случайный лес с подбором гиперпараметров</a></span></li></ul></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Изменение-порога" data-toc-modified-id="Изменение-порога-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Изменение порога</a></span></li><li><span><a href="#Балансировка-весов" data-toc-modified-id="Балансировка-весов-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Балансировка весов</a></span></li><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Увеличение выборки</a></span></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Уменьшение выборки</a></span></li></ul></li><li><span><a href="#ВЫВОДЫ:" data-toc-modified-id="ВЫВОДЫ:-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>ВЫВОДЫ:</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

## Подготовка данных

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.preprocessing import OrdinalEncoder, StandardScaler
from sklearn.metrics import mean_squared_error, recall_score, precision_score, f1_score, roc_auc_score
import numpy as np
from sklearn.utils import shuffle
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV


In [2]:
df = pd.read_csv('/datasets/Churn.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             9091 non-null float64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


В датасете присутствуют столбцы с типом object. Это необходимо исправить.

In [4]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [5]:
# наверное, было необязательно переименовывать название колонок
df = df.rename(columns={'RowNumber': 'row_number', 'CustomerId': 'customer_id', 'CreditScore':'credit_score', 
                        'NumOfProducts': 'bank_products', 'HasCrCard': 'credit_card', 
                        'IsActiveMember': 'active_member', 'EstimatedSalary': 'estimated_salary'})

In [6]:
# приводим все названия к нижнему регистру
df = df.rename(columns=str.lower)

In [7]:
df.isna().sum()

row_number            0
customer_id           0
surname               0
credit_score          0
geography             0
gender                0
age                   0
tenure              909
balance               0
bank_products         0
credit_card           0
active_member         0
estimated_salary      0
exited                0
dtype: int64

In [8]:
df['tenure'] = df['tenure'].fillna(0)

В данных обнаружились пропуски. Так как пропусков более 10%, удалить не получится, поэтому заменяем их нулями

In [9]:
# удаляем ненужные колонки со значениями, которые невозможно привести в численный вид
df = df.drop(['surname', 'customer_id'], axis=1)

In [10]:
# приводим категориальные признаки в численный вид методом ОНЕ, подходящим для целей классификации
df_ohe = pd.get_dummies(df, drop_first=True)


In [11]:
df_ohe.head() 

row_number  credit_score  age  tenure    balance  bank_products  \
0           1           619   42     2.0       0.00              1   
1           2           608   41     1.0   83807.86              1   
2           3           502   42     8.0  159660.80              3   
3           4           699   39     1.0       0.00              2   
4           5           850   43     2.0  125510.82              1   

   credit_card  active_member  estimated_salary  exited  geography_Germany  \
0            1              1         101348.88       1                  0   
1            0              1         112542.58       0                  0   
2            1              0         113931.57       1                  0   
3            0              0          93826.63       0                  0   
4            1              1          79084.10       0                  0   

   geography_Spain  gender_Male  
0                0            0  
1                1            0  
2                0            0  
3                0            0  
4                1            0

In [12]:
df_ohe.corr()

row_number  credit_score       age    tenure   balance  \
row_number           1.000000      0.005840  0.000783  0.000596 -0.009067   
credit_score         0.005840      1.000000 -0.003965  0.003087  0.006268   
age                  0.000783     -0.003965  1.000000 -0.007368  0.028308   
tenure               0.000596      0.003087 -0.007368  1.000000 -0.005821   
balance             -0.009067      0.006268  0.028308 -0.005821  1.000000   
bank_products        0.007246      0.012238 -0.030680  0.010106 -0.304180   
credit_card          0.000599     -0.005458 -0.011721  0.021387 -0.014858   
active_member        0.012044      0.025651  0.085472 -0.025856 -0.010084   
estimated_salary    -0.005988     -0.001384 -0.007201  0.011225  0.012797   
exited              -0.016571     -0.027094  0.285323 -0.013319  0.118533   
geography_Germany   -0.000044      0.005538  0.046897  0.001676  0.401110   
geography_Spain     -0.009905      0.004780 -0.001685 -0.000745 -0.134892   
gender_Male          0.018196     -0.002857 -0.027544  0.014889  0.012087   

                   bank_products  credit_card  active_member  \
row_number              0.007246     0.000599       0.012044   
credit_score            0.012238    -0.005458       0.025651   
age                    -0.030680    -0.011721       0.085472   
tenure                  0.010106     0.021387      -0.025856   
balance                -0.304180    -0.014858      -0.010084   
bank_products           1.000000     0.003183       0.009612   
credit_card             0.003183     1.000000      -0.011866   
active_member           0.009612    -0.011866       1.000000   
estimated_salary        0.014204    -0.009933      -0.011421   
exited                 -0.047820    -0.007138      -0.156128   
geography_Germany      -0.010419     0.010577      -0.020486   
geography_Spain         0.009039    -0.013480       0.016732   
gender_Male            -0.021859     0.005766       0.022544   

                   estimated_salary    exited  geography_Germany  \
row_number                -0.005988 -0.016571          -0.000044   
credit_score              -0.001384 -0.027094           0.005538   
age                       -0.007201  0.285323           0.046897   
tenure                     0.011225 -0.013319           0.001676   
balance                    0.012797  0.118533           0.401110   
bank_products              0.014204 -0.047820          -0.010419   
credit_card               -0.009933 -0.007138           0.010577   
active_member             -0.011421 -0.156128          -0.020486   
estimated_salary           1.000000  0.012097           0.010297   
exited                     0.012097  1.000000           0.173488   
geography_Germany          0.010297  0.173488           1.000000   
geography_Spain           -0.006482 -0.052667          -0.332084   
gender_Male               -0.008112 -0.106512          -0.024628   

                   geography_Spain  gender_Male  
row_number               -0.009905     0.018196  
credit_score              0.004780    -0.002857  
age                      -0.001685    -0.027544  
tenure                   -0.000745     0.014889  
balance                  -0.134892     0.012087  
bank_products             0.009039    -0.021859  
credit_card              -0.013480     0.005766  
active_member             0.016732     0.022544  
estimated_salary         -0.006482    -0.008112  
exited                   -0.052667    -0.106512  
geography_Germany        -0.332084    -0.024628  
geography_Spain           1.000000     0.016889  
gender_Male               0.016889     1.000000

Взаимосвязанных признаков не выявлено.

In [13]:
# отделяем целевой признак
features = df_ohe.drop(['exited'], axis=1)
target = df_ohe['exited']

In [14]:
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.20, random_state=12345
)

In [15]:
features_train, features_test, target_train, target_test = train_test_split(
    features_train, target_train, test_size=0.25, random_state=12345
)

In [16]:
print(features_train.shape, target_train.shape)

(6000, 12) (6000,)


In [17]:
print(features_valid.shape, target_valid.shape)

(2000, 12) (2000,)


In [18]:
print(features_test.shape, target_test.shape)

(2000, 12) (2000,)


Делим выборку на обучающую, валидационную и тестовую

In [19]:
 # преобразуем масштаб методом стандартизации
scaler = StandardScaler()
numeric =['credit_score', 'age', 'tenure', 'balance', 'estimated_salary', 'bank_products']
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/opt/conda/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


*** Выводы:**
* размер датасета достаточен для исследования.
* в датасете присутствуют столбцы с неподходящим типом.
* пропуски заполнены 0, так как не представляется возможным заполнить их оригинальными значениями.
* удалены признаки, которые невозможно привести в численный вид.
* категориальные признаки приведены в численный вид методом ОНЕ.
* признаки не взаимосвязаны.
* датасет разделен на выборки в соотношении 60:20:20.
* признаки стандартизированы.

## Исследование задачи

### Проверка на сбалансированность данных

In [20]:
df_ohe['exited'].value_counts(normalize=True)

0    0.7963
1    0.2037
Name: exited, dtype: float64

Подсчет классов в целевом признаке выявляет проблему несбалансированных данных. Объектов класса "1" меньше в четыре раза.

### Построение моделей на несбалансированных данных

#### Логистическая регрессия

In [21]:
model_logreg = LogisticRegression(random_state=12345, solver='liblinear')
model_logreg.fit(features_train, target_train)
predicted_valid = model_logreg.predict(features_valid)
probabilities_valid = model_logreg.predict_proba(features_valid)
probabilities_valid = probabilities_valid[:,1]
print('F1:', f1_score(target_valid, predicted_valid))
print('AUC-Roc:', roc_auc_score(target_valid, probabilities_valid))

F1: 0.28419182948490235
AUC-Roc: 0.7567365570346196


Модель логистической регрессии на несбалансированных данных показала очень низкое качество - она почти равна константной модели.

#### Модель случайного леса с подбором параметров

In [22]:
%%time
#param_grid =  {'n_estimators': range (10, 100, 10),
              #'max_depth': range (1,13, 2),
              #'min_samples_leaf': range (1,8),
              #'min_samples_split': range (2,10,2)}

model_forest = RandomForestClassifier(random_state=12345, max_depth=11, min_samples_leaf=1, min_samples_split=8,
                               n_estimators=90)
#model = GridSearchCV(estimator = clf, param_grid=param_grid,
                               #cv = 4, n_jobs = -1)
model_forest.fit(features_train,target_train)

predictionforest = model_forest.predict(features_valid)
probabilities_valid = model_forest.predict_proba(features_valid)
probabilities_valid= probabilities_valid[:,1]
print('f1:', f1_score(target_valid,predictionforest))
print('AUC-ROC:', roc_auc_score(target_valid, probabilities_valid))

f1: 0.548238897396631
AUC-ROC: 0.8659507407644518
CPU times: user 1.06 s, sys: 3.64 ms, total: 1.07 s
Wall time: 1.07 s


Данная модель значительно улучшила метрики качества.

Выводы:
* Подсчет классов в целевом признаке выявляет проблему несбалансированных данных.
* Модель логистической регрессии на несбалансированных данных оказалась равной константной модели.
* Модель случайного леса с подбором параметров значительно лучше по результату. Но скорость обучения сильно уменьшилась.

## Борьба с дисбалансом

### Изменение порога

#### Логистическая регрессия

In [24]:
model_log = LogisticRegression(random_state=12345, solver='liblinear')
model_log.fit(features_train, target_train)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=12345, solver='liblinear', tol=0.0001,
                   verbose=0, warm_start=False)

In [25]:
probabilities_valid = model_log.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:,1]

In [26]:
for threshold in np.arange(0.12, 0.3, 0.02):
    predicted_valid = probabilities_one_valid > threshold
    precision = precision_score(target_valid, predicted_valid) 
    recall = recall_score(target_valid, predicted_valid)
    f1= f1_score(target_valid, predicted_valid)
    #auc_roc= roc_auc_score(target_valid, predicted_valid)

    print("'Порог' = {:.2F}  |  'Точность' = {:.3F}  |  'Полнота' = {:.3F}  | 'F1' = {:.3F}"
          .format(threshold, precision, recall, f1), '|', 
          "'AUC_ROC' = {:.3F}".format(roc_auc_score(target_valid, probabilities_one_valid)))

'Порог' = 0.12  |  'Точность' = 0.294  |  'Полнота' = 0.883  | 'F1' = 0.441 | 'AUC_ROC' = 0.757
'Порог' = 0.14  |  'Точность' = 0.312  |  'Полнота' = 0.834  | 'F1' = 0.454 | 'AUC_ROC' = 0.757
'Порог' = 0.16  |  'Точность' = 0.330  |  'Полнота' = 0.796  | 'F1' = 0.467 | 'AUC_ROC' = 0.757
'Порог' = 0.18  |  'Точность' = 0.352  |  'Полнота' = 0.754  | 'F1' = 0.480 | 'AUC_ROC' = 0.757
'Порог' = 0.20  |  'Точность' = 0.371  |  'Полнота' = 0.714  | 'F1' = 0.488 | 'AUC_ROC' = 0.757
'Порог' = 0.22  |  'Точность' = 0.395  |  'Полнота' = 0.681  | 'F1' = 0.500 | 'AUC_ROC' = 0.757
'Порог' = 0.24  |  'Точность' = 0.418  |  'Полнота' = 0.644  | 'F1' = 0.507 | 'AUC_ROC' = 0.757
'Порог' = 0.26  |  'Точность' = 0.428  |  'Полнота' = 0.604  | 'F1' = 0.501 | 'AUC_ROC' = 0.757
'Порог' = 0.28  |  'Точность' = 0.441  |  'Полнота' = 0.564  | 'F1' = 0.495 | 'AUC_ROC' = 0.757


Модель логистической регрессии неплохие метрики качества при пороге 0.24: 'F1' = 0.507, auc_roc 0.757

#### Модель случайного леса с подбором параметров

In [27]:
model_forest = RandomForestClassifier(random_state=12345, max_depth=11, min_samples_leaf=1, min_samples_split=8,
                                      n_estimators=90)
model_forest.fit(features_train, target_train)
probabilities_valid_forest = model_forest.predict_proba(features_valid)
probabilities_valid_forest = probabilities_valid_forest[:,1]

In [28]:
for threshold in np.arange(0.1, 0.3, 0.02):
    predicted_valid = probabilities_valid_forest > threshold
    precision = precision_score(target_valid, predicted_valid) 
    recall = recall_score(target_valid, predicted_valid)
    f1= f1_score(target_valid, predicted_valid)

    print("'Порог' = {:.2F}  |  'Точность' = {:.3F}  |  'Полнота' = {:.3F}  | 'F1' = {:.3F}"
          .format(threshold, precision, recall, f1), '|', 
          "'AUC_ROC' = {:.3F}".format(roc_auc_score(target_valid, probabilities_valid_forest)))

'Порог' = 0.10  |  'Точность' = 0.337  |  'Полнота' = 0.918  | 'F1' = 0.493 | 'AUC_ROC' = 0.866
'Порог' = 0.12  |  'Точность' = 0.369  |  'Полнота' = 0.895  | 'F1' = 0.523 | 'AUC_ROC' = 0.866
'Порог' = 0.14  |  'Точность' = 0.403  |  'Полнота' = 0.867  | 'F1' = 0.550 | 'AUC_ROC' = 0.866
'Порог' = 0.16  |  'Точность' = 0.443  |  'Полнота' = 0.852  | 'F1' = 0.583 | 'AUC_ROC' = 0.866
'Порог' = 0.18  |  'Точность' = 0.466  |  'Полнота' = 0.824  | 'F1' = 0.595 | 'AUC_ROC' = 0.866
'Порог' = 0.20  |  'Точность' = 0.498  |  'Полнота' = 0.810  | 'F1' = 0.617 | 'AUC_ROC' = 0.866
'Порог' = 0.22  |  'Точность' = 0.529  |  'Полнота' = 0.773  | 'F1' = 0.628 | 'AUC_ROC' = 0.866
'Порог' = 0.24  |  'Точность' = 0.560  |  'Полнота' = 0.733  | 'F1' = 0.635 | 'AUC_ROC' = 0.866
'Порог' = 0.26  |  'Точность' = 0.575  |  'Полнота' = 0.700  | 'F1' = 0.631 | 'AUC_ROC' = 0.866
'Порог' = 0.28  |  'Точность' = 0.605  |  'Полнота' = 0.674  | 'F1' = 0.638 | 'AUC_ROC' = 0.866


Данная модель показыват хорошую прибавку в качестве при порог 0.28: 'F1' = 0.638, auc_roc 0.866

### Балансировка весов класса

#### Логистическая регрессия

In [29]:
model_balanced = LogisticRegression(class_weight='balanced', solver='liblinear', random_state=12345)
model_balanced.fit(features_train, target_train)
predicted_balanced = model_balanced.predict(features_valid)
probabilities_valid = model_balanced.predict_proba(features_valid)
probabilities_valid = probabilities_valid[:,1]
print('F1:', f1_score(target_valid, predicted_balanced))
print('AUC-ROC:', roc_auc_score(target_valid, probabilities_valid))

F1: 0.5029436501261564
AUC-ROC: 0.7644590878570013


Модель логистической регрессии результат почти не улучшила по сравнению с балансировкой весов. 

#### Случайный лес с подбором гиперпараметров

In [30]:
%%time

model_forest_balanced = RandomForestClassifier(random_state=12345, class_weight='balanced', max_depth=11, min_samples_leaf=1, min_samples_split=8,
                                      n_estimators=90)
model_forest_balanced.fit(features_train,target_train)

predictionforest = model_forest_balanced.predict(features_valid)
probabilities_valid = model_forest_balanced.predict_proba(features_valid)
probabilities_valid = probabilities_valid[:,1]
print('f1:', f1_score(target_valid,predictionforest))
print('AUC-ROC:', roc_auc_score(target_valid, probabilities_valid))

f1: 0.6198547215496367
AUC-ROC: 0.868477275332715
CPU times: user 1.01 s, sys: 38.9 ms, total: 1.05 s
Wall time: 1.13 s


При балансировки весов классов модель случайного леса слегка потеряла в качестве по сравнению с моделью с измененным порогом.

### Увеличение выборки

In [31]:
def upsample(features, target, repeat):
    '''функция делит признаки на классы и увеличивает класс "1" в заданное количество раз(repeat)'''
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    features_upsampled, target_upsampled = shuffle(
    features_upsampled, target_upsampled, random_state=12345)
    return features_upsampled, target_upsampled    

#### Логистическая регрессия

In [32]:
features_upsampled, target_upsampled = upsample(features_train, target_train, 4)
model_ups =  LogisticRegression(solver='liblinear', random_state=12345)
model_ups.fit(features_upsampled, target_upsampled)
predict_ups = model_ups.predict(features_valid)
probabilities_valid = model_ups.predict_proba(features_valid)
probabilities_valid = probabilities_valid[:,1]
print('F1:', f1_score(target_valid, predict_ups))
print('AUC-ROC:', roc_auc_score(target_valid, probabilities_valid))

F1: 0.5020781379883624
AUC-ROC: 0.7644873755156916


Увеличение выборки тоже не помогло увеличить метрики качесва модели

#### Случайный лес с подбором гиперпараметров

In [33]:
model_forest_up = RandomForestClassifier(random_state=12345, max_depth=11, min_samples_leaf=1, min_samples_split=8,
                                      n_estimators=90)
model_forest_up.fit(features_upsampled, target_upsampled)
probabilities_valid = model_forest_up.predict_proba(features_valid)
probabilities_valid = probabilities_valid[:,1]
predictionforest = model_forest_up.predict(features_valid)
print('f1:', f1_score(target_valid,predictionforest))
print('AUC-ROC:', roc_auc_score(target_valid, probabilities_valid))

f1: 0.634254143646409
AUC-ROC: 0.8661949079236709


### Уменьшение выборки

In [34]:
def downsample(features, target, fraction):
    '''функция делит признаки на классы и уменьшает класс "0" и отбрасывает часть отрицательных объектов'''
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    features_downsampled = pd.concat(
    [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
    [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    features_downsampled, target_downsampled = shuffle(
    features_downsampled, target_downsampled, random_state=12345)
    return features_downsampled, target_downsampled

#### Логистическая регрессия

In [35]:
features_downsampled, target_downsampled = downsample(features_train, target_train, 0.3)
model_log_down =  LogisticRegression(solver='liblinear', random_state=12345)
model_log_down.fit(features_downsampled, target_downsampled)
predict_valid = model_log_down.predict(features_valid)
probabilities_valid = model_log_down.predict_proba(features_valid)
probabilities_valid = probabilities_valid[:,1]
print('F1:', f1_score(target_valid, predict_valid))
print('AUC-ROC:', roc_auc_score(target_valid, probabilities_valid))

F1: 0.5122171945701357
AUC-ROC: 0.7661444367852713


#### Случайный лес с подбором гиперпараметров

In [36]:
model_forest_down = RandomForestClassifier(random_state=12345, max_depth=11, min_samples_leaf=1, min_samples_split=8,
                                      n_estimators=90)
model_forest_down.fit(features_downsampled, target_downsampled)

predictionforest = model_forest_down.predict(features_valid)
probabilities_valid = model_forest_down.predict_proba(features_valid)
probabilities_valid = probabilities_valid[:,1]
print('f1:', f1_score(target_valid,predictionforest))
print('AUC-ROC:', roc_auc_score(target_valid, probabilities_valid))

f1: 0.625250501002004
AUC-ROC: 0.8662738156031747


In [37]:
display(pd.DataFrame([['Logistic Regression, F1', 0.284, 0.507, 0.503, 0.502, 0.512],
                  ['Logistic Regression, AUC-ROC', 0.757, 0.757, 0.764, 0.764, 0.766],
                 ['Random Forest, F1', 0.548, 0.628, 0.620, 0.634, 0.625],
                 ['Random Forest, AUC-ROC', 0.866, 0.866, 0.868, 0.866, 0.866]],
                  columns=['Model name','Несбалансированные данные', 'Изменение порога', 'Балансировка весов',
                           'Увеличение выборки', 'Уменьшение выборки']))

Model name  Несбалансированные данные  Изменение порога  \
0       Logistic Regression, F1                      0.284             0.507   
1  Logistic Regression, AUC-ROC                      0.757             0.757   
2             Random Forest, F1                      0.548             0.628   
3        Random Forest, AUC-ROC                      0.866             0.866   

   Балансировка весов  Увеличение выборки  Уменьшение выборки  
0               0.503               0.502               0.512  
1               0.764               0.764               0.766  
2               0.620               0.634               0.625  
3               0.868               0.866               0.866

При применени различных свредств для улучшения качества модели лучший результат показывает модель случайного леса с подбором параметров при изменении порога.
Модель логистической регрессии показывает примерно одинаковые результаты.

## Тестирование модели

In [38]:
# объединение обучающей и валидационной выборок
features_concat = pd.concat([features_train] + [features_valid])
target_concat = pd.concat([target_train] + [target_valid])
features_concat = shuffle(features_concat, random_state=12345)
target_concat = shuffle(target_concat, random_state=12345)


### Изменение порога

In [39]:
model_log = LogisticRegression(random_state=12345, solver='liblinear')
model_log.fit(features_concat, target_concat)
probabilities_test = model_log.predict_proba(features_test)
probabilities_one_test = probabilities_test[:,1]

In [40]:
for threshold in np.arange(0.15, 0.3, 0.02):
    predicted_test = probabilities_one_test > threshold
    precision = precision_score(target_test, predicted_test) 
    recall = recall_score(target_test, predicted_test)
    f1= f1_score(target_test, predicted_test)

    print("'Порог' = {:.2F}  |  'Точность' = {:.3F}  |  'Полнота' = {:.3F}  | 'F1' = {:.3F}"
          .format(threshold, precision, recall, f1), '|', 
          "'AUC_ROC' = {:.3F}".format(roc_auc_score(target_test, probabilities_one_test)))

'Порог' = 0.15  |  'Точность' = 0.299  |  'Полнота' = 0.834  | 'F1' = 0.440 | 'AUC_ROC' = 0.768
'Порог' = 0.17  |  'Точность' = 0.326  |  'Полнота' = 0.788  | 'F1' = 0.461 | 'AUC_ROC' = 0.768
'Порог' = 0.19  |  'Точность' = 0.341  |  'Полнота' = 0.734  | 'F1' = 0.466 | 'AUC_ROC' = 0.768
'Порог' = 0.21  |  'Точность' = 0.358  |  'Полнота' = 0.693  | 'F1' = 0.473 | 'AUC_ROC' = 0.768
'Порог' = 0.23  |  'Точность' = 0.384  |  'Полнота' = 0.647  | 'F1' = 0.482 | 'AUC_ROC' = 0.768
'Порог' = 0.25  |  'Точность' = 0.398  |  'Полнота' = 0.601  | 'F1' = 0.479 | 'AUC_ROC' = 0.768
'Порог' = 0.27  |  'Точность' = 0.426  |  'Полнота' = 0.570  | 'F1' = 0.488 | 'AUC_ROC' = 0.768
'Порог' = 0.29  |  'Точность' = 0.444  |  'Полнота' = 0.529  | 'F1' = 0.483 | 'AUC_ROC' = 0.768


У модели логистической регрессии лучший результат при пороге 0.27

In [41]:
model_forest_weigth = RandomForestClassifier(random_state=12345, max_depth=11, min_samples_leaf=1,
                                             min_samples_split=8, n_estimators=90)
model_forest_weigth.fit(features_concat, target_concat)
probabilities_test = model_forest_weigth.predict_proba(features_test)
probabilities_one_test = probabilities_test[:,1]


In [42]:
for threshold in np.arange(0.15, 0.4, 0.02):
    predicted_test = probabilities_one_test > threshold
    precision = precision_score(target_test, predicted_test) 
    recall = recall_score(target_test, predicted_test)
    f1= f1_score(target_test, predicted_test)

    print("'Порог' = {:.2F}  |  'Точность' = {:.3F}  |  'Полнота' = {:.3F}  | 'F1' = {:.3F}"
          .format(threshold, precision, recall, f1), '|', 
           "'AUC_ROC' = {:.3F}".format(roc_auc_score(target_test, probabilities_one_test)))

'Порог' = 0.15  |  'Точность' = 0.379  |  'Полнота' = 0.818  | 'F1' = 0.518 | 'AUC_ROC' = 0.851
'Порог' = 0.17  |  'Точность' = 0.411  |  'Полнота' = 0.788  | 'F1' = 0.540 | 'AUC_ROC' = 0.851
'Порог' = 0.19  |  'Точность' = 0.437  |  'Полнота' = 0.754  | 'F1' = 0.553 | 'AUC_ROC' = 0.851
'Порог' = 0.21  |  'Точность' = 0.458  |  'Полнота' = 0.731  | 'F1' = 0.564 | 'AUC_ROC' = 0.851
'Порог' = 0.23  |  'Точность' = 0.483  |  'Полнота' = 0.711  | 'F1' = 0.575 | 'AUC_ROC' = 0.851
'Порог' = 0.25  |  'Точность' = 0.502  |  'Полнота' = 0.678  | 'F1' = 0.577 | 'AUC_ROC' = 0.851
'Порог' = 0.27  |  'Точность' = 0.528  |  'Полнота' = 0.647  | 'F1' = 0.582 | 'AUC_ROC' = 0.851
'Порог' = 0.29  |  'Точность' = 0.549  |  'Полнота' = 0.634  | 'F1' = 0.588 | 'AUC_ROC' = 0.851
'Порог' = 0.31  |  'Точность' = 0.600  |  'Полнота' = 0.624  | 'F1' = 0.612 | 'AUC_ROC' = 0.851
'Порог' = 0.33  |  'Точность' = 0.616  |  'Полнота' = 0.596  | 'F1' = 0.606 | 'AUC_ROC' = 0.851
'Порог' = 0.35  |  'Точность' = 0.636  |

Случайный лес - при пороге 0.31

### Балансировка весов

In [43]:
model_balanced = LogisticRegression(class_weight='balanced', solver='liblinear', random_state=12345)
model_balanced.fit(features_concat, target_concat)
predicted_balance = model_balanced.predict(features_test)
probabilities_test = model_balanced.predict_proba(features_test)
probabilities_test = probabilities_test[:,1]
print('F1:', f1_score(target_test, predicted_balance))
print('AUC-ROC:', roc_auc_score(target_test, probabilities_test))

F1: 0.483754512635379
AUC-ROC: 0.7731828159696338


In [44]:
model_forest_balanced = RandomForestClassifier(random_state=12345, class_weight='balanced', max_depth=11, min_samples_leaf=1, min_samples_split=8,
                               n_estimators=90)
model_forest_balanced.fit(features_concat,target_concat)

predicted_test = model_forest_balanced.predict(features_test)
probabilities_test = model_forest_balanced.predict_proba(features_test)
probabilities_test = probabilities_test[:,1]
print('f1:', f1_score(target_test,predicted_test))
print('AUC-ROC:', roc_auc_score(target_test,probabilities_test))

f1: 0.588380716934487
AUC-ROC: 0.8529340236107954


### Увеличение выборки

In [45]:
features_concat_up, target_concat_up = upsample(features_concat, target_concat, 4)
model_log.fit(features_concat_up, target_concat_up)
predict_test = model_log.predict(features_test)
probabilities_test = model_log.predict_proba(features_test)
probabilities_test = probabilities_test[:,1]
print('F1:', f1_score(target_test, predict_test))
print('AUC-ROC:', roc_auc_score(target_test, probabilities_test))

F1: 0.484472049689441
AUC-ROC: 0.7732034797868131


In [46]:
model_forest = RandomForestClassifier(random_state=12345, max_depth=11, min_samples_leaf=1, min_samples_split=8,
                               n_estimators=90)
model_forest.fit(features_concat_up, target_concat_up)
predict_test = model_forest.predict(features_test)
probabilities_test = model_forest.predict_proba(features_test)
probabilities_test = probabilities_test[:,1]
print('F1:', f1_score(target_test, predict_test))
print('AUC-ROC:', roc_auc_score(target_test, probabilities_test))

F1: 0.5862457722660653
AUC-ROC: 0.8517879765195456


### Уменьшение выборки

In [47]:
features_concat_down, target_concat_down = downsample(features_concat, target_concat, 0.3)
model_log_down =  LogisticRegression(solver='liblinear', random_state=12345)
model_log_down.fit(features_concat_down, target_concat_down)
predict_test = model_log_down.predict(features_test)
probabilities_test = model_log_down.predict_proba(features_test)
probabilities_test = probabilities_test[:,1]
print('F1:', f1_score(target_test, predict_test))
print('AUC-ROC:', roc_auc_score(target_test, probabilities_test))

F1: 0.47722772277227726
AUC-ROC: 0.7715328896440896


In [48]:
model_forest = RandomForestClassifier(random_state=12345, max_depth=11, min_samples_leaf=1, min_samples_split=8,
                               n_estimators=90)
model_forest.fit(features_concat_down, target_concat_down)
predict_test = model_forest.predict(features_test)
probabilities_test = model_forest.predict_proba(features_test)
probabilities_test = probabilities_test[:,1]
print('F1:', f1_score(target_test, predict_test))
print('AUC-ROC:', roc_auc_score(target_test, probabilities_test))

F1: 0.5822510822510822
AUC-ROC: 0.8494211746903209


In [49]:
print('Сравнение показателей при тестировании')
display(pd.DataFrame([['Logistic Regression, F1', 0.488, 0.484, 0.484, 0.477],
                     ['Logistic Regression, AUC-ROC', 0.768, 0.773, 0.773, 0.772],
                    ['Random Forest, F1', 0.612, 0.588, 0.586, 0.582],
                    ['Random Forest, AUC-ROC', 0.851, 0.853, 0.851, 0.849]],
                    columns=['Model name','Изменение порога', 'Балансировка весов',
                           'Увеличение выборки', 'Уменьшение выборки']))

Сравнение показателей при тестировании


Model name  Изменение порога  Балансировка весов  \
0       Logistic Regression, F1             0.488               0.484   
1  Logistic Regression, AUC-ROC             0.768               0.773   
2             Random Forest, F1             0.612               0.588   
3        Random Forest, AUC-ROC             0.851               0.853   

   Увеличение выборки  Уменьшение выборки  
0               0.484               0.477  
1               0.773               0.772  
2               0.586               0.582  
3               0.851               0.849

При тестировании моделей лучшие показатели обеих моделей оказались: у логистической регрессии при балансировке весов и увеличении выборки, у случайного леса при изменении порога.

## ВЫВОДЫ:
* размер датасета достаточен для исследования.
* в датасете исправлены типы данных.
* пропуски заполнены.
* удалены признаки, которые невозможно привести в численный вид.
* категориальные признаки приведены в численный вид методом ОНЕ.
* датасет разделен на выборки в соотношении 60:20:20.
* признаки стандартизированы.
* построены модели логистической регрессии и случайного леса с подбором параметров.
* проведены исследования для улучшения показателей качества моделей.
* при обучении и тестировании лучшей оказалась модель случайного леса с изменением порога.